src/dtree.ml

```diff
@@ -0,0 +1,398 @@
+    birth to children nodes.  The label on the edge between a node and its
+    child represents the term matched to generate the next pattern matrix (the
+    one of the child node); and is therefore one of the terms in the column of
+    the pattern matrix whose index is the label of the node.  The first [d]
+    edge is due to the initialization matrix (with {!recfield:origin} [=]
+    {!cons:Init}). *)
+let to_dot : string -> t -> unit = fun fname tree ->
+  let module F = Format in
+  let module P = Print in
+  let ochan = open_out (fname ^ ".gv") in
+  let ppf = F.formatter_of_out_channel ochan in
+  let nodecount = ref 0 in
+  F.fprintf ppf "graph {@[<v>" ;
+  let pp_opterm : term option pp = fun oc teo -> match teo with
+    | Some(t) -> P.pp oc t
+    | None    -> F.fprintf oc "d" in
+  let rec write_tree : int -> term option -> t -> unit =
+    fun father_l swon tree ->
+  (* We cannot use iter since we need the father to be passed. *)
+    match tree with
+    | Leaf(a)  ->
+      incr nodecount ;
+      F.fprintf ppf "@ %d [label=\"" !nodecount ;
+      let _, acte = Bindlib.unmbind a in
+      P.pp ppf acte ; F.fprintf ppf "\"]" ;
+      F.fprintf ppf "@ %d -- %d [label=\"" father_l !nodecount ;
+      pp_opterm ppf swon ; F.fprintf ppf "\"];"
+    | Node(ndata) ->
+      let { swap = swa ; children = ch ; _ } = ndata in
+      incr nodecount ;
+      let tag = !nodecount in
+      F.fprintf ppf "@ %d [label=\"" tag ;
+      F.fprintf ppf "%d" (match swa with None -> 0 | Some(i) -> i) ;
+      F.fprintf ppf "\"]" ;
+      F.fprintf ppf "@ %d -- %d [label=\"" father_l tag ;
+      pp_opterm ppf swon ; F.fprintf ppf "\"];" ;
+      List.iter (fun (s, e) -> write_tree tag s e) ch ;
+    | Fail        ->
+      incr nodecount ;
+      F.fprintf ppf "@ %d -- %d [label=\"%s\"];" father_l !nodecount "f"
+  in
+  begin
+    match tree with
+    (* First step must be done to avoid drawing a top node. *)
+    | Node({ swap = _ ; children = ch ; _ }) ->
+      List.iter (fun (sw, c) -> write_tree 0 sw c) ch
+    | Leaf(_)                                -> ()
+    | _                                      -> assert false
+  end ;
+  F.fprintf ppf "@.}@\n@?" ;
+  close_out ochan
+
+(** Pattern matrices is a way to encode a pattern matching problem.  A line is
+    a candidate instance of the values matched.  Each line is a pattern having
+    an action attached.  This module provides functions to operate on these
+    matrices. *)
+module Pattmat =
+struct
+  (** Type used to describe a line of a matrix (either a column or a row). *)
+  type line = term list
+
+  (** A redefinition of the rule type. *)
+  type rule = { lhs : line
+              (** Left hand side of a rule. *)
+              ; rhs : action
+              (** Right hand side of a rule. *)
+              ; env : term_env array
+              (** Used to pattern match variables. *)}
+
+  (** The core data, contains the rewrite rules. *)
+  type matrix = rule list
+
```

Comment:
Possible improvement for later: use arrays instead of lists.

Reply:
I don't think that it is relevant as we almost always process all the rules

Reply:
But don't you use List.nth many times?

Reply:
it was the case when the rules were themselves lists (of terms), but they are now arrays